influxdata/tokenizer.go

```diff
@@ -5,93 +5,195 @@ import (
 outer:
 	for {
-		if !t.ensure(n + 1) {
+		if !t.ensure(1) {
 			break
 		}
-		buf := t.buf[t.r+n:]
+		buf := t.buf[t.r1:]
 		for i, c := range buf {
 			if !set.get(c) {
-				n += i
+				t.r1 += i
 				break outer
 			}
 		}
-		n += len(buf)
+		t.r1 += len(buf)
+	}
+	return t.buf[t.r0+start : t.r1]
+}
+
+// takeEsc is like take except that escaped characters also count as
+// part of the set. The escapeTable determines which characters
+// can be escaped.
+// It returns the unescaped string (unless t.skipping is true, in which
+// case it doesn't need to go to the trouble of unescaping it).
+func (t *Tokenizer) takeEsc(set *byteSet, escapeTable *[256]byte) []byte {
+	// start holds the offset from r0 of the start of the taken slice.
+	// Note that we can't use t.r1 directly, because the offsets can change
+	// when the buffer is grown.
+	start := t.r1 - t.r0
+
+	// startUnesc holds the offset from t0 of the start of the most recent
+	// unescaped segment.
+	startUnesc := start
+
+	// startEsc holds the index into r.escBuf of the start of the escape buffer.
+	startEsc := len(t.escBuf)
+outer:
+	for {
+		if !t.ensure(1) {
+			break
+		}
+		buf := t.buf[t.r1:]
+		for i := 0; i < len(buf); i++ {
+			c := buf[i]
+			if c != '\\' {
+				if !set.get(c) {
+					// We've found the end, so we're done here.
+					t.r1 += i
+					break outer
+				}
+				continue
+			}
+			if i+1 >= len(buf) {
+				// Not enough room in the buffer. Try reading more so that
+				// we can see the next character.
+				if !t.ensure(i + 2) {
+					// No character to escape, so leave the \ intact.
+					t.r1 = len(t.buf)
+					break outer
+				}
+				// Note that t.ensure can change t.buf, so we need to
+				// update buf to point to the correct buffer.
+				buf = t.buf[t.r1:]
+			}
+			replc := escapeTable[buf[i+1]]
+			if replc == 0 {
+				// The backslash doesn't precede a value escaped
+				// character, so it stays intact.
+				continue
+			}
+			if !t.skipping {
+				t.escBuf = append(t.escBuf, t.buf[t.r0+startUnesc:t.r1+i]...)
+				t.escBuf = append(t.escBuf, replc)
+				startUnesc = t.r1 - t.r0 + i + 2
+			}
+			i++
+		}
+		// We've consumed all the bytes in the buffer. Now continue
+		// the loop, trying to acquire more.
+		t.r1 += len(buf)
+	}
+	if len(t.escBuf) > startEsc {
+		// We've got an unescaped result: append any remaining unescaped bytes
+		// and return the relevant portion of the escape buffer.
+		t.escBuf = append(t.escBuf, t.buf[startUnesc+t.r0:t.r1]...)
+		return t.escBuf[startEsc:]
 	}
-	return t.buf[t.r+from : t.r+n]
+	return t.buf[t.r0+start : t.r1]
 }
 
 // at returns the byte at i bytes after the current read position.
-// It assumes that the byte has already been read (use t.ensure to check
-// whether it's there).
+// It assumes that the index has already been ensured.
+// If there's no byte there, it returns zero.
 func (t *Tokenizer) at(i int) byte {
-	return t.buf[t.r+i]
+	if t.r1+i < len(t.buf) {
+		return t.buf[t.r1+i]
+	}
+	return 0
```

Comment:
This line isn't covered by tests.

Reply:
I don't mind in this case, because the logic is so simple, and the internal tests are about providing initial confidence. It'll probably be covered later when the surrounding code uses this primitive a little more.